File: fujsen/src/CoreOperators/JsonOperator.php
<?php
/**
 * JSON Operator
 * 
 * JSON manipulation operator for encoding, decoding, and processing
 * JSON data with validation and formatting options.
 * 
 * @package TuskPHP\CoreOperators
 * @version 2.0.0
 */

namespace TuskPHP\CoreOperators;

/**
 * JSON Operator
 * 
 * Provides JSON manipulation capabilities including encoding, decoding,
 * validation, and formatting with various options.
 */
class JsonOperator extends BaseOperator
{
    public function __construct()
    {
        $this->version = '2.0.0';
        $this->requiredFields = ['data'];
        $this->optionalFields = [
            'action', 'pretty', 'validate', 'schema', 'options'
        ];
        
        $this->defaultConfig = [
            'action' => 'encode',
            'pretty' => false,
            'validate' => false,
            'options' => []
        ];
    }
    
    public function getName(): string
    {
        return 'json';
    }
    
    protected function getDescription(): string
    {
        return 'JSON manipulation operator for encoding, decoding, and processing JSON data with validation and formatting options';
    }
    
    protected function getExamples(): array
    {
        return [
            'encode' => '@json({data: {name: "John", age: 30}, pretty: true})',
            'decode' => '@json({action: "decode", data: "{\"name\":\"John\"}"})',
            'validate' => '@json({action: "validate", data: json_string, schema: schema_object})',
            'merge' => '@json({action: "merge", data: [obj1, obj2, obj3]})',
            'extract' => '@json({action: "extract", data: json_object, path: "users[0].name"})',
            'simple' => '@json({name: "value"})'
        ];
    }
    
    protected function getErrorCodes(): array
    {
        return array_merge(parent::getErrorCodes(), [
            'JSON_ENCODE_ERROR' => 'JSON encoding failed',
            'JSON_DECODE_ERROR' => 'JSON decoding failed',
            'INVALID_SCHEMA' => 'Invalid JSON schema',
            'VALIDATION_FAILED' => 'JSON validation failed',
            'INVALID_PATH' => 'Invalid JSON path'
        ]);
    }
    
    /**
     * Execute JSON operator
     */
    protected function executeOperator(array $config, array $context): mixed
    {
        $action = $config['action'];
        $data = $this->resolveVariable($config['data'], $context);
        
        switch ($action) {
            case 'encode':
                return $this->encodeJson($data, $config);
            case 'decode':
                return $this->decodeJson($data, $config);
            case 'validate':
                return $this->validateJson($data, $config, $context);
            case 'merge':
                return $this->mergeJson($data, $config);
            case 'extract':
                return $this->extractFromJson($data, $config, $context);
            case 'transform':
                return $this->transformJson($data, $config, $context);
            default:
                // Default to encode
                return $this->encodeJson($data, $config);
        }
    }
    
    /**
     * Encode data to JSON
     */
    private function encodeJson(mixed $data, array $config): string
    {
        $pretty = $config['pretty'];
        $options = $config['options'] ?? [];
        
        // Build JSON options
        $jsonOptions = 0;
        
        if ($pretty) {
            $jsonOptions |= JSON_PRETTY_PRINT;
        }
        
        if (isset($options['unescaped_slashes']) && $options['unescaped_slashes']) {
            $jsonOptions |= JSON_UNESCAPED_SLASHES;
        }
        
        if (isset($options['unicode_escape']) && $options['unicode_escape']) {
            $jsonOptions |= JSON_UNESCAPED_UNICODE;
        }
        
        if (isset($options['force_object']) && $options['force_object']) {
            $jsonOptions |= JSON_FORCE_OBJECT;
        }
        
        $result = json_encode($data, $jsonOptions);
        
        if ($result === false) {
            $error = json_last_error_msg();
            throw new \RuntimeException("JSON encoding failed: {$error}");
        }
        
        $this->log('info', 'JSON encoded', [
            'pretty' => $pretty,
            'options' => $jsonOptions,
            'length' => strlen($result)
        ]);
        
        return $result;
    }
    
    /**
     * Decode JSON string
     */
    private function decodeJson(string $data, array $config): mixed
    {
        $assoc = $config['assoc'] ?? true;
        $depth = $config['depth'] ?? 512;
        $options = $config['options'] ?? 0;
        
        $result = json_decode($data, $assoc, $depth, $options);
        
        if ($result === null && json_last_error() !== JSON_ERROR_NONE) {
            $error = json_last_error_msg();
            throw new \RuntimeException("JSON decoding failed: {$error}");
        }
        
        $this->log('info', 'JSON decoded', [
            'assoc' => $assoc,
            'depth' => $depth,
            'type' => gettype($result)
        ]);
        
        return $result;
    }
    
    /**
     * Validate JSON against schema
     */
    private function validateJson(mixed $data, array $config, array $context): bool
    {
        $schema = $this->resolveVariable($config['schema'], $context);
        
        if (!$schema) {
            // Simple validation - just check if it's valid JSON
            if (is_string($data)) {
                json_decode($data);
                return json_last_error() === JSON_ERROR_NONE;
            }
            return true;
        }
        
        // Basic schema validation (simplified)
        return $this->validateAgainstSchema($data, $schema);
    }
    
    /**
     * Merge multiple JSON objects
     */
    private function mergeJson(array $data, array $config): array
    {
        $strategy = $config['strategy'] ?? 'recursive';
        
        if (empty($data)) {
            return [];
        }
        
        $result = $data[0];
        
        for ($i = 1; $i < count($data); $i++) {
            $result = $this->mergeObjects($result, $data[$i], $strategy);
        }
        
        $this->log('info', 'JSON merged', [
            'count' => count($data),
            'strategy' => $strategy
        ]);
        
        return $result;
    }
    
    /**
     * Extract value from JSON using path
     */
    private function extractFromJson(mixed $data, array $config, array $context): mixed
    {
        $path = $this->resolveVariable($config['path'], $context);
        
        if (!$path) {
            return $data;
        }
        
        $result = $this->extractByPath($data, $path);
        
        $this->log('info', 'JSON extracted', [
            'path' => $path,
            'found' => $result !== null
        ]);
        
        return $result;
    }
    
    /**
     * Transform JSON data
     */
    private function transformJson(mixed $data, array $config, array $context): mixed
    {
        $transform = $this->resolveVariable($config['transform'], $context);
        
        if (!$transform || !is_callable($transform)) {
            return $data;
        }
        
        $result = call_user_func($transform, $data);
        
        $this->log('info', 'JSON transformed', [
            'transform' => gettype($transform)
        ]);
        
        return $result;
    }
    
    /**
     * Validate data against schema
     */
    private function validateAgainstSchema(mixed $data, array $schema): bool
    {
        // Basic schema validation (simplified implementation)
        foreach ($schema as $key => $rule) {
            if (!isset($data[$key])) {
                if (isset($rule['required']) && $rule['required']) {
                    return false;
                }
                continue;
            }
            
            $value = $data[$key];
            
            // Type validation
            if (isset($rule['type'])) {
                if (!$this->validateType($value, $rule['type'])) {
                    return false;
                }
            }
            
            // Pattern validation
            if (isset($rule['pattern']) && is_string($value)) {
                if (!preg_match($rule['pattern'], $value)) {
                    return false;
                }
            }
            
            // Range validation
            if (isset($rule['min']) && is_numeric($value)) {
                if ($value < $rule['min']) {
                    return false;
                }
            }
            
            if (isset($rule['max']) && is_numeric($value)) {
                if ($value > $rule['max']) {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    /**
     * Validate value type
     */
    private function validateType(mixed $value, string $type): bool
    {
        switch ($type) {
            case 'string':
                return is_string($value);
            case 'number':
                return is_numeric($value);
            case 'integer':
                return is_int($value);
            case 'boolean':
                return is_bool($value);
            case 'array':
                return is_array($value);
            case 'object':
                return is_object($value);
            case 'null':
                return $value === null;
            default:
                return true;
        }
    }
    
    /**
     * Merge two objects
     */
    private function mergeObjects(array $obj1, array $obj2, string $strategy): array
    {
        if ($strategy === 'replace') {
            return array_merge($obj1, $obj2);
        }
        
        // Recursive merge
        $result = $obj1;
        
        foreach ($obj2 as $key => $value) {
            if (isset($result[$key]) && is_array($result[$key]) && is_array($value)) {
                $result[$key] = $this->mergeObjects($result[$key], $value, $strategy);
            } else {
                $result[$key] = $value;
            }
        }
        
        return $result;
    }
    
    /**
     * Extract value by path
     */
    private function extractByPath(mixed $data, string $path): mixed
    {
        $parts = explode('.', $path);
        $current = $data;
        
        foreach ($parts as $part) {
            // Handle array access
            if (preg_match('/^(.+)\[(\d+)\]$/', $part, $matches)) {
                $key = $matches[1];
                $index = (int)$matches[2];
                
                if (!isset($current[$key]) || !is_array($current[$key])) {
                    return null;
                }
                
                if (!isset($current[$key][$index])) {
                    return null;
                }
                
                $current = $current[$key][$index];
            } else {
                if (!isset($current[$part])) {
                    return null;
                }
                
                $current = $current[$part];
            }
        }
        
        return $current;
    }
    
    /**
     * Custom validation
     */
    protected function customValidate(array $config): array
    {
        $errors = [];
        $warnings = [];
        
        // Validate action
        if (isset($config['action'])) {
            $validActions = ['encode', 'decode', 'validate', 'merge', 'extract', 'transform'];
            if (!in_array($config['action'], $validActions)) {
                $errors[] = "Invalid action: {$config['action']}";
            }
        }
        
        // Validate options
        if (isset($config['options']) && !is_array($config['options'])) {
            $errors[] = "Options must be an array";
        }
        
        return ['errors' => $errors, 'warnings' => $warnings];
    }
} 

File: sdk/csharp/csharp-sdk-package-20250721_013844/operators/Operators/DataProcessing/JsonOperator.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Text.Json;

namespace TuskLang.Operators.DataProcessing
{
    /// <summary>
    /// JSON Operator for TuskLang C# SDK
    /// 
    /// Provides JSON processing capabilities with support for:
    /// - JSON parsing and serialization
    /// - JSON validation and formatting
    /// - JSON path queries (JSONPath)
    /// - JSON schema validation
    /// - JSON transformation and manipulation
    /// - Pretty printing and minification
    /// 
    /// Usage:
    /// ```csharp
    /// // Parse JSON
    /// var result = @json({
    ///   action: "parse",
    ///   data: "{\"name\": \"John\", \"age\": 30}"
    /// })
    /// 
    /// // Query with JSONPath
    /// var result = @json({
    ///   action: "query",
    ///   data: jsonData,
    ///   path: "$.users[*].name"
    /// })
    /// ```
    /// </summary>
    public class JsonOperator : BaseOperator
    {
        public JsonOperator()
        {
            Version = "2.0.0";
            RequiredFields = new List<string> { "action" };
            OptionalFields = new List<string> 
            { 
                "data", "path", "schema", "format", "indent", "validate", 
                "transform", "merge", "diff", "patch", "timeout" 
            };
            
            DefaultConfig = new Dictionary<string, object>
            {
                ["format"] = "pretty",
                ["indent"] = 2,
                ["validate"] = true,
                ["timeout"] = 300
            };
        }
        
        public override string GetName() => "json";
        
        protected override string GetDescription() => "JSON processing operator for parsing, querying, and manipulating JSON data";
        
        protected override Dictionary<string, string> GetExamples()
        {
            return new Dictionary<string, string>
            {
                ["parse"] = "@json({action: \"parse\", data: \"{\\\"name\\\": \\\"John\\\"}\"})",
                ["stringify"] = "@json({action: \"stringify\", data: {name: \"John\", age: 30}})",
                ["query"] = "@json({action: \"query\", data: jsonData, path: \"$.users[*].name\"})",
                ["validate"] = "@json({action: \"validate\", data: jsonData, schema: schemaData})"
            };
        }
        
        protected override Dictionary<string, string> GetErrorCodes()
        {
            return new Dictionary<string, string>
            {
                ["INVALID_ACTION"] = "Invalid JSON action",
                ["INVALID_JSON"] = "Invalid JSON data",
                ["PARSE_ERROR"] = "JSON parsing error",
                ["QUERY_ERROR"] = "JSONPath query error",
                ["SCHEMA_VALIDATION_ERROR"] = "JSON schema validation failed",
                ["TRANSFORM_ERROR"] = "JSON transformation error",
                ["MERGE_ERROR"] = "JSON merge error",
                ["TIMEOUT_EXCEEDED"] = "JSON operation timeout exceeded"
            };
        }
        
        protected override async Task<object> ExecuteOperatorAsync(Dictionary<string, object> config, Dictionary<string, object> context)
        {
            var action = GetContextValue<string>(config, "action", "");
            var data = ResolveVariable(config.GetValueOrDefault("data"), context);
            var path = GetContextValue<string>(config, "path", "");
            var schema = ResolveVariable(config.GetValueOrDefault("schema"), context);
            var format = GetContextValue<string>(config, "format", "pretty");
            var indent = GetContextValue<int>(config, "indent", 2);
            var validate = GetContextValue<bool>(config, "validate", true);
            var transform = GetContextValue<string>(config, "transform", "");
            var merge = ResolveVariable(config.GetValueOrDefault("merge"), context);
            var diff = ResolveVariable(config.GetValueOrDefault("diff"), context);
            var patch = ResolveVariable(config.GetValueOrDefault("patch"), context);
            var timeout = GetContextValue<int>(config, "timeout", 300);
            
            if (string.IsNullOrEmpty(action))
                throw new ArgumentException("Action is required");
            
            var startTime = DateTime.UtcNow;
            
            try
            {
                // Check timeout
                if ((DateTime.UtcNow - startTime).TotalSeconds > timeout)
                {
                    throw new TimeoutException("JSON operation timeout exceeded");
                }
                
                switch (action.ToLower())
                {
                    case "parse":
                        return await ParseJsonAsync(data, validate);
                    
                    case "stringify":
                        return await StringifyJsonAsync(data, format, indent);
                    
                    case "query":
                        return await QueryJsonAsync(data, path);
                    
                    case "validate":
                        return await ValidateJsonAsync(data, schema);
                    
                    case "transform":
                        return await TransformJsonAsync(data, transform);
                    
                    case "merge":
                        return await MergeJsonAsync(data, merge);
                    
                    case "diff":
                        return await DiffJsonAsync(data, diff);
                    
                    case "patch":
                        return await PatchJsonAsync(data, patch);
                    
                    case "format":
                        return await FormatJsonAsync(data, format, indent);
                    
                    case "minify":
                        return await MinifyJsonAsync(data);
                    
                    case "flatten":
                        return await FlattenJsonAsync(data);
                    
                    case "unflatten":
                        return await UnflattenJsonAsync(data);
                    
                    default:
                        throw new ArgumentException($"Unknown JSON action: {action}");
                }
            }
            catch (Exception ex)
            {
                Log("error", "JSON operation failed", new Dictionary<string, object>
                {
                    ["action"] = action,
                    ["error"] = ex.Message
                });
                
                return new Dictionary<string, object>
                {
                    ["success"] = false,
                    ["error"] = ex.Message,
                    ["action"] = action
                };
            }
        }
        
        /// <summary>
        /// Parse JSON string to object
        /// </summary>
        private async Task<object> ParseJsonAsync(object data, bool validate)
        {
            if (data == null)
                throw new ArgumentException("Data is required for parsing");
            
            var jsonString = data.ToString();
            
            try
            {
                var options = new JsonSerializerOptions
                {
                    PropertyNameCaseInsensitive = true,
                    AllowTrailingCommas = true,
                    ReadCommentHandling = JsonCommentHandling.Skip
                };
                
                var result = JsonSerializer.Deserialize<object>(jsonString, options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["parsed"] = result,
                    ["valid"] = true
                };
            }
            catch (JsonException ex)
            {
                if (validate)
                {
                    throw new ArgumentException($"Invalid JSON: {ex.Message}");
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = false,
                    ["error"] = ex.Message,
                    ["valid"] = false
                };
            }
        }
        
        /// <summary>
        /// Convert object to JSON string
        /// </summary>
        private async Task<object> StringifyJsonAsync(object data, string format, int indent)
        {
            if (data == null)
                throw new ArgumentException("Data is required for stringification");
            
            try
            {
                var options = new JsonSerializerOptions
                {
                    WriteIndented = format == "pretty",
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                };
                
                var jsonString = JsonSerializer.Serialize(data, options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["stringified"] = jsonString,
                    ["format"] = format
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Stringification failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Query JSON with JSONPath
        /// </summary>
        private async Task<object> QueryJsonAsync(object data, string path)
        {
            if (data == null)
                throw new ArgumentException("Data is required for querying");
            
            if (string.IsNullOrEmpty(path))
                throw new ArgumentException("Path is required for querying");
            
            try
            {
                // Simplified JSONPath implementation
                // In a real implementation, you would use a proper JSONPath library
                var results = new List<object>();
                
                if (path == "$" || path == "$.*")
                {
                    results.Add(data);
                }
                else if (path.StartsWith("$."))
                {
                    var propertyPath = path.Substring(2);
                    var result = ExtractProperty(data, propertyPath);
                    if (result != null)
                    {
                        results.Add(result);
                    }
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["results"] = results,
                    ["count"] = results.Count,
                    ["path"] = path
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"JSONPath query failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Validate JSON against schema
        /// </summary>
        private async Task<object> ValidateJsonAsync(object data, object schema)
        {
            if (data == null)
                throw new ArgumentException("Data is required for validation");
            
            // Simplified validation
            // In a real implementation, you would use a proper JSON schema validator
            var isValid = true;
            var errors = new List<string>();
            
            if (schema != null)
            {
                // Basic schema validation logic would go here
                // For now, just check if data is not null
                if (data == null)
                {
                    isValid = false;
                    errors.Add("Data cannot be null");
                }
            }
            
            return new Dictionary<string, object>
            {
                ["success"] = true,
                ["valid"] = isValid,
                ["errors"] = errors,
                ["error_count"] = errors.Count
            };
        }
        
        /// <summary>
        /// Transform JSON data
        /// </summary>
        private async Task<object> TransformJsonAsync(object data, string transform)
        {
            if (data == null)
                throw new ArgumentException("Data is required for transformation");
            
            if (string.IsNullOrEmpty(transform))
                throw new ArgumentException("Transform expression is required");
            
            try
            {
                // Simplified transformation
                // In a real implementation, you would parse and execute the transform expression
                var transformed = data; // Placeholder for actual transformation
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["transformed"] = transformed,
                    ["transform"] = transform
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Transformation failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Merge JSON objects
        /// </summary>
        private async Task<object> MergeJsonAsync(object data, object merge)
        {
            if (data == null)
                throw new ArgumentException("Data is required for merging");
            
            if (merge == null)
                throw new ArgumentException("Merge data is required");
            
            try
            {
                // Simplified merge
                // In a real implementation, you would implement deep merge logic
                var merged = new Dictionary<string, object>();
                
                if (data is Dictionary<string, object> dataDict)
                {
                    foreach (var kvp in dataDict)
                    {
                        merged[kvp.Key] = kvp.Value;
                    }
                }
                
                if (merge is Dictionary<string, object> mergeDict)
                {
                    foreach (var kvp in mergeDict)
                    {
                        merged[kvp.Key] = kvp.Value;
                    }
                }
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["merged"] = merged
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Merge failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Diff JSON objects
        /// </summary>
        private async Task<object> DiffJsonAsync(object data, object diff)
        {
            if (data == null)
                throw new ArgumentException("Data is required for diffing");
            
            if (diff == null)
                throw new ArgumentException("Diff data is required");
            
            try
            {
                // Simplified diff
                // In a real implementation, you would implement proper diff logic
                var differences = new List<object>();
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["differences"] = differences,
                    ["has_differences"] = differences.Count > 0
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Diff failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Patch JSON object
        /// </summary>
        private async Task<object> PatchJsonAsync(object data, object patch)
        {
            if (data == null)
                throw new ArgumentException("Data is required for patching");
            
            if (patch == null)
                throw new ArgumentException("Patch data is required");
            
            try
            {
                // Simplified patch
                // In a real implementation, you would implement proper patch logic
                var patched = data; // Placeholder for actual patching
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["patched"] = patched
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Patch failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Format JSON
        /// </summary>
        private async Task<object> FormatJsonAsync(object data, string format, int indent)
        {
            if (data == null)
                throw new ArgumentException("Data is required for formatting");
            
            try
            {
                var options = new JsonSerializerOptions
                {
                    WriteIndented = format == "pretty",
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                };
                
                var formatted = JsonSerializer.Serialize(data, options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["formatted"] = formatted,
                    ["format"] = format
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Formatting failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Minify JSON
        /// </summary>
        private async Task<object> MinifyJsonAsync(object data)
        {
            if (data == null)
                throw new ArgumentException("Data is required for minification");
            
            try
            {
                var options = new JsonSerializerOptions
                {
                    WriteIndented = false
                };
                
                var minified = JsonSerializer.Serialize(data, options);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["minified"] = minified
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Minification failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Flatten JSON object
        /// </summary>
        private async Task<object> FlattenJsonAsync(object data)
        {
            if (data == null)
                throw new ArgumentException("Data is required for flattening");
            
            try
            {
                var flattened = FlattenObject(data, "");
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["flattened"] = flattened
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Flattening failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Unflatten JSON object
        /// </summary>
        private async Task<object> UnflattenJsonAsync(object data)
        {
            if (data == null)
                throw new ArgumentException("Data is required for unflattening");
            
            try
            {
                var unflattened = UnflattenObject(data);
                
                return new Dictionary<string, object>
                {
                    ["success"] = true,
                    ["unflattened"] = unflattened
                };
            }
            catch (Exception ex)
            {
                throw new ArgumentException($"Unflattening failed: {ex.Message}");
            }
        }
        
        /// <summary>
        /// Extract property from object using path
        /// </summary>
        private object ExtractProperty(object data, string path)
        {
            if (data == null || string.IsNullOrEmpty(path))
                return null;
            
            var parts = path.Split('.');
            var current = data;
            
            foreach (var part in parts)
            {
                if (current is Dictionary<string, object> dict)
                {
                    if (!dict.TryGetValue(part, out current))
                    {
                        return null;
                    }
                }
                else
                {
                    return null;
                }
            }
            
            return current;
        }
        
        /// <summary>
        /// Flatten object to key-value pairs
        /// </summary>
        private Dictionary<string, object> FlattenObject(object data, string prefix)
        {
            var result = new Dictionary<string, object>();
            
            if (data is Dictionary<string, object> dict)
            {
                foreach (var kvp in dict)
                {
                    var key = string.IsNullOrEmpty(prefix) ? kvp.Key : $"{prefix}.{kvp.Key}";
                    
                    if (kvp.Value is Dictionary<string, object> nestedDict)
                    {
                        var nested = FlattenObject(kvp.Value, key);
                        foreach (var nestedKvp in nested)
                        {
                            result[nestedKvp.Key] = nestedKvp.Value;
                        }
                    }
                    else
                    {
                        result[key] = kvp.Value;
                    }
                }
            }
            
            return result;
        }
        
        /// <summary>
        /// Unflatten object from key-value pairs
        /// </summary>
        private object UnflattenObject(object data)
        {
            if (data is Dictionary<string, object> dict)
            {
                var result = new Dictionary<string, object>();
                
                foreach (var kvp in dict)
                {
                    var parts = kvp.Key.Split('.');
                    var current = result;
                    
                    for (int i = 0; i < parts.Length - 1; i++)
                    {
                        var part = parts[i];
                        if (!current.ContainsKey(part))
                        {
                            current[part] = new Dictionary<string, object>();
                        }
                        current = (Dictionary<string, object>)current[part];
                    }
                    
                    current[parts[parts.Length - 1]] = kvp.Value;
                }
                
                return result;
            }
            
            return data;
        }
        
        protected override ValidationResult CustomValidate(Dictionary<string, object> config)
        {
            var result = new ValidationResult();
            
            if (!config.ContainsKey("action"))
            {
                result.Errors.Add("Action is required");
            }
            
            var action = GetContextValue<string>(config, "action", "");
            var validActions = new[] { "parse", "stringify", "query", "validate", "transform", "merge", "diff", "patch", "format", "minify", "flatten", "unflatten" };
            
            if (!string.IsNullOrEmpty(action) && !Array.Exists(validActions, a => a == action.ToLower()))
            {
                result.Errors.Add($"Invalid action: {action}. Valid actions are: {string.Join(", ", validActions)}");
            }
            
            if (config.TryGetValue("timeout", out var timeout) && timeout is int timeoutValue && timeoutValue <= 0)
            {
                result.Errors.Add("Timeout must be positive");
            }
            
            if (config.TryGetValue("indent", out var indent) && indent is int indentValue && indentValue < 0)
            {
                result.Errors.Add("Indent must be non-negative");
            }
            
            return result;
        }
    }
} 